realm/realm-library/src/objectServer/java/io/realm/SyncSession.java

```diff
@@ -114,5 +130,185 @@ protected void finalize() throws Throwable {
+                    // This is acceptable.
+                    networkListener = new NetworkStateReceiver.ConnectionListener() {
+                        @Override
+                        public void onChange(boolean connectionAvailable) {
+                            if (connectionAvailable) {
+                                if (!onGoingAccessTokenQuery.getAndSet(true)) {
+                                    authenticateRealm(authServer);
+                                }
+                                NetworkStateReceiver.removeListener(this);
+                            }
+                        }
+                    };
+                    NetworkStateReceiver.addListener(networkListener);
+                }
+            }
+        }
+        return null;
+    }
+
+    // Authenticate by getting access tokens for the specific Realm
+    private void authenticateRealm(final AuthenticationServer authServer) {
+        if (networkRequest != null) {
+            networkRequest.cancel();
+        }
+        clearScheduledAccessTokenRefresh();
+
+        // Authenticate in a background thread. This allows incremental backoff and retries in a safe manner.
+        Future<?> task = SyncManager.NETWORK_POOL_EXECUTOR.submit(new ExponentialBackoffTask<AuthenticateResponse>() {
+            @Override
+            protected AuthenticateResponse execute() {
+                if (!isClosed && !Thread.currentThread().isInterrupted()) {
+                    return authServer.loginToRealm(
+                            getUser().getAccessToken(),//refresh token in fact
+                            configuration.getServerUrl(),
+                            getUser().getSyncUser().getAuthenticationUrl()
+                    );
+                }
+                return null;
+            }
+
+            @Override
+            protected void onSuccess(AuthenticateResponse response) {
+                RealmLog.debug("Session[%s]: Access token acquired", configuration.getPath());
+                if (!isClosed && !Thread.currentThread().isInterrupted()) {
+                    ObjectServerUser.AccessDescription desc = new ObjectServerUser.AccessDescription(
+                            response.getAccessToken(),
+                            configuration.getPath(),
+                            configuration.shouldDeleteRealmOnLogout()
+                    );
+                    getUser().getSyncUser().addRealm(configuration.getServerUrl(), desc);
+                    // schedule a token refresh before it expires
+                    if (nativeRefreshAccessToken(configuration.getPath(), getUser().getSyncUser().getAccessToken(configuration.getServerUrl()).value(), configuration.getServerUrl().toString())) {
+                        scheduleRefreshAccessToken(authServer, response.getAccessToken().expiresMs());
+
+                    } else {
+                        // token not applied, no refresh will be scheduled
```

Comment:
we interface with the ObjectStore using the `realm_path`, so if for some use case (see below) the ObjectStore session is no longer found `nativeRefreshAccessToken` and we couldn't apply the `access_token` we just fetched, then stop the refreshing scheduling.

One use case, is the user open a Realm, we start refreshing, by the time the network query returns, the user have already closed the last instance of this Realm & the ObjectStore destroys the session, so when calling `get_session` in `nativeRefreshAccessToken` you'll get nothing.